sklearn/tests/test_pipeline.py

```diff
@@ -1814,6 +1821,140 @@ def test_pipeline_inverse_transform_Xt_deprecation():
+        assert registry
+        for estimator in registry:
+            for method in methods:
+                check_recorded_metadata(
+                    estimator,
+                    method=method,
+                    parent=method,
+                    **metadata,
+                )
+
+    X = np.array([[1, 2], [3, 4]])
+    y = np.array([0, 1])
+    sample_weight = np.array([[1, 2]])
+    other_weights = np.array([[30, 40]])
+    metadata = np.array([[100, 200]])
+
+    pipe, registry_1, registry_2, registry_3, registry_4 = get_pipeline()
+    pipe.fit(
+        X,
+        y,
+        sample_weight=sample_weight,
+        other_weights=other_weights,
+        metadata=metadata,
+    )
+
+    check_metadata(
+        registry_1, ["fit", "transform"], sample_weight=sample_weight, metadata=metadata
+    )
+    check_metadata(registry_2, ["fit", "transform"])
+    check_metadata(
+        registry_3,
+        ["fit", "transform"],
+        sample_weight=sample_weight + 2,
+        metadata=metadata,
+    )
+    check_metadata(
+        registry_4,
+        method.split("_"),  # ["fit", "transform"] if "fit_transform", ["fit"] otherwise
+        sample_weight=other_weights + 3,
+        metadata=metadata,
+    )
+
+
+def test_transform_input_no_slep6():
+    """Make sure the right error is raised if slep6 is not enabled."""
+    X = np.array([[1, 2], [3, 4]])
+    y = np.array([0, 1])
+    msg = "The `transform_input` parameter can only be set if metadata"
+    with pytest.raises(ValueError, match=msg):
+        make_pipeline(DummyTransf(), transform_input=["blah"]).fit(X, y)
+
+
+@pytest.mark.usefixtures("enable_slep006")
+def test_transform_tuple_input():
+    """Test that if metadata is a tuple of arrays, both arrays are transformed."""
+
+    class Estimator(ClassifierMixin, BaseEstimator):
+        def fit(self, X, y, X_val=None, y_val=None):
+            assert isinstance(X_val, tuple)
+            assert isinstance(y_val, tuple)
+            # Here we make sure that each X_val is transformed by the transformer
+            assert_array_equal(X_val[0], np.array([[2, 3]]))
+            assert_array_equal(y_val[0], np.array([0, 1]))
+            assert_array_equal(X_val[1], np.array([[11, 12]]))
+            assert_array_equal(y_val[1], np.array([1, 2]))
+            return self
+
+    class Transformer(TransformerMixin, BaseEstimator):
+        def fit(self, X, y):
+            return self
+
+        def transform(self, X):
+            return X + 1
+
+    X = np.array([[1, 2]])
+    y = np.array([0, 1])
+    X_val0 = np.array([[1, 2]])
+    y_val0 = np.array([0, 1])
+    X_val1 = np.array([[10, 11]])
+    y_val1 = np.array([1, 2])
+    pipe = Pipeline(
+        [
+            ("transformer", Transformer()),
+            ("estimator", Estimator().set_fit_request(X_val=True, y_val=True)),
+        ],
+        transform_input=["X_val"],
+    )
+    pipe.fit(X, y, X_val=(X_val0, X_val1), y_val=(y_val0, y_val1))
```

Comment:
@lorentzenchr @StrikerRUS @trivialfis the tuple pattern is now present and tested here.

It would be lovely if you could test this with a few cases.

Reply:
If no one else gets to it sooner, I'd be happy to try testing this (for both `lightgbm` and `xgboost`).

I could probably get to that some time in the next week... but I won't be offended if you say "thanks but that's too long to wait, we're going to merge this soon".

Reply:
@jameslamb yeah go for it. Thanks.

Reply:
I built from this branch tonight and ran `lightgbm` and `xgboost`'s scikit-learn tests.

<details><summary>code to do that (click me)</summary>

I've omitted some details (like installing dependencies) here intentionally, to focus on the main steps. Can provide a more complete setup if you want.

```shell
# build and install scikit-learn from this branch
pip uninstall --yes scikit-learn
git clone \
    --recursive \
    --branch pipeline/transform \
    git@github.com:adrinjalali/scikit-learn.git \
    /tmp/scikit-learn

pushd /tmp/scikit-learn
pip install --no-deps .
popd

# xgboost
cd ~/repos/xgboost/python-package
pip install --no-deps .
cd ..
pytest \
    --disable-warnings \
    --tb=short \
    -rs \
    ./tests/python/test_with_sklearn.py

# lightgbm
cd ~/repos/LightGBM
sh build-python.sh install bdist_wheel
pytest tests/python_package_test/test_sklearn.py
```

</details>

I did find that `xgboost` will need some other changes to adjust to things coming in `scikit-learn` 1.6 like `lightgbm` did (https://github.com/dmlc/xgboost/issues/10896), but didn't see any test failures related to this branch's changes 🎉 

In the next couple days, I'll try something similar to https://github.com/scikit-learn/scikit-learn/pull/28901#issuecomment-2325946444 with those libraries. @adrinjalali if you have time, it'd be helpful if you could update this branch to latest `main` before I do that.

Reply:
the branch is updated @jameslamb

Reply:
@adrinjalali @lorentzenchr I had a chance to look at the discussion and changes in this PR more closely... as implemented, this is not going to work for `lightgbm` and `xgboost`.

Those projects' estimators don't accept separate tuples of arrays for `X` and `y` like this:

```python
est.fit(X, y, X_val=(X_val0, X_val1), y_val=(y_val0, y_val1))
```

Their `fit()` methods have a single argument `eval_set` which accepts a Python list of tuples of `X, y` pairs, like this:

```python
est.fit(X, y, eval_set=[(X_val0, y_val0), (X_val1, y_val1)])
```

https://github.com/microsoft/LightGBM/blob/dc0ed538aa09b755ba4a293dc6f344da51674260/python-package/lightgbm/sklearn.py#L341-L342

https://github.com/dmlc/xgboost/blob/9e81fb39f69e83cfd5a9800d205793f89cd416d4/python-package/xgboost/sklearn.py#L1090-L1093

Reply:
Yes, hence the "if" statement in this comment: https://github.com/scikit-learn/scikit-learn/pull/28901#issuecomment-2325946444

Those two libraries would need to change the API to enable this pattern.

Reply:
I see. When I read that comment before, I didn't understand that by "if the API changes" you meant "if `xgboost` / `lightgbm` change their APIs".

Changing straight to that would be a significant breaking change for `xgboost` / `lightgbm`. Validation sets are crucially important for GBDTs as an enabler of early stopping, so I suspect there are many many existing training pipelines out in the world using `xgboost` / `lightgbm` scikit-learn estimators and passing `eval_set`.

On the other hand, I do understand that it'd be valuable to transform validation sets in a `Pipeline`.

If we (`xgboost` / `lightgbm`) *were* to pursue this, I'd recommend this path that should minimize disruption for users:

* **adding** new arguments `eval_X` and `eval_y` to `fit()` methods that accept iterables of inputs as in the example above
* raising `ValueError` if `eval_set` is non-empty and either of `eval_X` / `eval_y` is non-empty (to avoid the complexity of having to merge arguments provided both ways)
* raising a deprecation warning if `eval_set` is non-empty, encouraging the use of `eval_X` and `eval_y` instead
* after a long deprecation period (like a year+), removing `eval_set`

cc @StrikerRUS @jmoralez @trivialfis @hcho3 do you have input here? Is it worth taking on this plan as feature requests in `xgboost` / `lightgbm`? If so, I can write up issues.

Reply:
@adrinjalali 
May I wonder, what was the reason behind implementing new API with pattern of separate tuples instead of adopting existing in XGBoost/LightGBM one of list of tuples?

Reply:
The API that already exists in xgboost/lightgbm, includes two data that are different in nature: X and y. X needs to be transformed in the pipeline up to the step where it's needed, while y shouldn't be touched.

Here, the user tells the pipeline what to transform, in the form of naming the argument to be transformed (e.g. `X_val`). If we keep the xgboost/lightgbm API, then the user would need to tell the pipeline that "the first member of the tuple in each of member inside the tuple/list given under an argument, should be transformed", in order to only transform the X part in a `[(X_val1, y_val1), (X_val2, y_val2)]`. And that's not something we can develop a nice API about. Decoupling `X_val` from `y_val` so that they're not given under the same input argument, makes it possible to simply say the contents of which argument should be transformed.

Reply:
> If we keep the xgboost/lightgbm API, then the user would need to tell the pipeline that "the first member of the tuple in each of member inside the tuple/list given under an argument, should be transformed"

That wouldn't *have to* be something users specify. Just as you've special-cased tuple arguments here, it seems to me that `scikit-learn` could safely special-case list-of-tuple arguments to assume they're `(X, y)` data.

> *... something we can develop a nice API about*

As far as I can tell, `HistGradientBoostingClassifier` doesn't yet allow passing validation sets (just a `validation_fraction` which randomly selects a subset of rows):

* `__init__()`: https://scikit-learn.org/dev/modules/generated/sklearn.ensemble.HistGradientBoostingClassifier.html#sklearn.ensemble.HistGradientBoostingClassifier
* `fit()`: https://scikit-learn.org/dev/modules/generated/sklearn.ensemble.HistGradientBoostingClassifier.html#sklearn.ensemble.HistGradientBoostingClassifier.fit

So there's time to co-design an API for this that works for both `scikit-learn`'s own estimators and `xgboost` / `lightgbm` scikit-learn-compatible estimators.

**option 1:** Match `lightgbm` / `xgboost`'s existing API. This would mean this validation set transformation is available to users of those projects immediately once `scikit-learn` 1.6 is released.

**option 2: Introducing this new API and then asking `lightgbm` / `xgboost` to adjust to it. This would some development work will be required in those libraries, and for all users of those libraries (assuming `lightgbm` / `xgboost` pursue deprecating the `eval_set` argument to not have to maintain 2 separate code paths for the same purpose). So very unlikely that this PR's functionality would be usable with those libraries when `scikit-learn` 1.6 is released.

Honestly... I am ok Option 2. I do think the separate `eval_X` and `eval_y` arguments is a better API. In `lightgbm`, we already have many other arguments to `fit()` which are "an iterable where the `i`th element corresponds to the `i`th validation set"

* `eval_sample_weight`
* `eval_class_weight`
* `eval_init_score`

docs: https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMClassifier.html#lightgbm.LGBMClassifier.fit

similar in XGBoost: https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBClassifier.fit

But we'd have to preserve the `eval_set` (list of `(X, y)` tuples) API for a **long** deprecation cycle :/

Reply:
The issue I see with the combination of keeping the `[(X,y), ...]` pattern is that the special case would need to skip transforming some of the given argument, and transform the other parts. The special case of tuple in this PR, only makes sure everything in that argument is transformed and nothing's skipped. I'd be less comfortable special casing something to skip a part of the input.

Reply:
@StrikerRUS , yes I added it exclusively for xgboost/lightgbm. We haven't had plans to have this API inside sklearn itself.

Reply:
A bit more context: One of the motivations for metadata routing was to support this use case of passing `X_val` and `y_val` to estimators that support early stopping. We now have the routing in place, 2 things are missing: 1) add `X_val`, `y_val` to the signature of `fit` in those estimators and 2) enable to transform `X_val` along the pipeline before it is passed to the estimator consuming it.
This PR is about figuring out 2) and Adrin added tuples in https://github.com/scikit-learn/scikit-learn/pull/28901/commits/cdaf20fa5d30013b8af2bf720cb7eed18cdea5fd to get closer to XGBoost/LightGBM API.